shared/dataflow/codeql/dataflow/internal/DataFlowImpl.qll

```diff
@@ -3511,6 +3511,19 @@ module MakeImpl<LocationSig Location, InputSig<Location> Lang> {
              * included in the module `PathGraph`.
              */
             predicate flowPath(PathNode source, PathNode sink) {
+              (
+                // When there are both sources and sinks in the diff range,
+                // diff-informed dataflow falls back to computing all paths without
+                // any filtering. To prevent significant alert flip-flopping due to
+                // minor code changes triggering the fallback, we consistently apply
+                // source-or-sink filtering here to ensure that we return the same
+                // paths regardless of whether the fallback is triggered.
+                if Config::observeDiffInformedIncrementalMode()
+                then
+                  AlertFiltering::filterByLocation(source.getLocation()) or
+                  AlertFiltering::filterByLocation(sink.getLocation())
```

Comment:
This will add a lot of inlined code and redo the `filterByLocation` that was already done in `isFilteredSource`/`isFilteredSink`. On the other hand we don't really want to expose those two predicates beyond the `SourceSinkFiltering` module, as they can be confused with `isRelevantSource`/`isRelevantSink`. So instead I'd suggest adding an inlined predicate to the `SourceSinkFiltering` module that performs this check.
```
      bindingset[source, sink]
      pragma[inline_late]
      predicate isRelevantSourceSinkPair(Node source, Node sink) {
        isFilteredSource(source) or
        isFilteredSink(sink)
      }
```
and change this like so:
```suggestion
                  isRelevantSourceSinkPair(source.getNode(), sink.getNode())
```

Reply:
(With this change, `module AlertFiltering` can remain private)